### PR TITLE
[#1646] Add documentation how to implement a third party protocol adapter

### PR DESCRIPTION
**Pull request by ko-j**
As discussed in #1478, the developer documentation is extended by an example to implement the third party adapter using adapted code of the hono cli client.

### CODE REVIEW

File: site/documentation/content/dev-guide/third_party_adapter.md [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+{{< figure src="../third_party_adapter.png" >}}  
+
+Devices send telemetry/event messages via their own proprietary protocols to the *Third Party Adapters* gateway,
+which send the messages via AMQP to the Hono AMQP adapter.
+
+A prerequisite for using *Third Party Adapters* is that devices are registered in Hono's Device Registry
+and are configured in such a way that the *Third Party Adapters* can act on their behalf i.e.
+with a device registration "via" property containing the gateway id of the *Third Party Adapters*.
+
```
Comment (sophokles73) [inline site/documentation/content/dev-guide/third_party_adapter.md]: FMPOV this should go into a separate document under *concepts*. I can imagine something like *Connecting Devices* in which we describe the different options for how devices connect to Hono, e.g. directly or via a gateway. There we could then also introduce the notion of what you call here *Third Party Adapter* but which we actually call a *Protocol Gateway* as it is not a protocol adapter but instead is just a standard *gateway* which is potentially able to scale out horizontally.

**Comment by b-abel** [comment]
> @b-abel @calohmn How does this relate to the work you have been doing so far regarding a client for implementing protocol gateways?

I'm in the process to provide such a client as a PR. The idea is to have it based on the code that is already in the client module but to adapt it so that it can be used to connect to the AMQP protocol adapter. The current classes in the client module cannot be used, mainly because the links that the AMQP adapter uses southbound are different from those it opens northbound.
The reason for using the existing client code is that it addresses a number of problems, e.g. regarding connection loss, which a Proton based solution must solve itself.
While the focus is on a protocol gateway scenario, I think it could be seen as a general client for the AMQP adapter that could be used in other use cases as well.

I hope to make it available as PR in the next few days. It's bad timing that we seem to be working on similar topics at the same time. The reason I'm doing this now is that I'm in contact with solution developers that are currently starting to work on a protocol gateway and have a very ambitious schedule. My goal is that some kind of example code for a protocol gateway will be contributed to Hono.

File: site/documentation/content/dev-guide/third_party_adapter.md (hunk 1 above)
Comment (ko-j) [reply site/documentation/content/dev-guide/third_party_adapter.md]: That sounds good. I took the naming from the issue, but find *Protocol Gateway* also better describing the module.

**Comment by ko-j** [comment]
@b-abel 
We can make a proper *gateway protocol* library draft for issue #1478 later.
I would suggest that for now I use the simple snippets from the `third_party_adapter.md` to explain the basic concepts of a *gateway protocol adapter* (f.k.a. *third party protocol adapter* )

In summary, I would now
- move the [`third_party_adapter.md`](https://github.com/eclipse/hono/pull/1724/files/57fee7a208067213837e73b62dea762bb238ea6d#diff-088cdd579ca75d56e4c683fd5143d237) to concepts
  - briefly explaining the options to connect southbound to Hono
  - replacing code snippets with links to *example* module
- move current [contents of *example*](https://github.com/eclipse/hono/tree/master/example) into a subfolder
  _Is this referenced from anywhere else but [dev-guide/java_client_consumer.md#L16-L31](https://github.com/bosch-io/hono/blame/d6e75b0f3aff288b20c1cc076bdb63492f35338e/site/documentation/content/dev-guide/java_client_consumer.md#L16-L31) ?_
- refactor snippets from [`third_party_adapter.md`](https://github.com/eclipse/hono/pull/1724/files/57fee7a208067213837e73b62dea762bb238ea6d#diff-088cdd579ca75d56e4c683fd5143d237) into code in a *example* subfolder

**Comment by sophokles73** [comment]
@ko-j any news on this one?

File: example/protocol-adapter-example/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+# set credential secret for device
+export MY_PWD="dummyDevicePassword"
+echo "MY_PWD=\"${MY_PWD}\""
+curl -i -X PUT -H "content-type: application/json" --data-binary '[{
+  "type": "hashed-password",
+  "auth-id": "'$MY_DEVICE'",
+  "secrets": [{ "pwd-plain": "'$MY_PWD'" }]
+}]' http://$REGISTRY_IP:28080/v1/credentials/$MY_TENANT/$MY_DEVICE
+
```
Comment (sophokles73) [inline example/protocol-adapter-example/README.md]: why not add this script to the repository?

File: example/protocol-adapter-example/pom.xml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
```
Comment (sophokles73) [inline example/protocol-adapter-example/pom.xml]: please add the copyright/license header

File: example/protocol-adapter-example/pom.xml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.springframework.boot</groupId>
+                <artifactId>spring-boot-maven-plugin</artifactId>
+            </plugin>
+        </plugins>
+    </build>
+
+</project>
```
Comment (sophokles73) [inline example/protocol-adapter-example/pom.xml]: missing EOL

File: example/protocol-adapter-example/pom.xml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>org.eclipse.hono.example.protocoladapter</groupId>
+    <version>0.0.1-SNAPSHOT</version>
+    <artifactId>example-protocoladapter</artifactId>
+    <name>protocol adapter example</name>
+    <description>A simple example to ilustrate how a protocol adapter could be implemented using classes from the Hono CLI module</description>
+
```
Comment (sophokles73) [inline example/protocol-adapter-example/pom.xml]: please define the parent and only set the artifactId here

File: example/protocol-adapter-example/pom.xml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    <properties>
+        <java.version>8</java.version>
```
Comment (sophokles73) [inline example/protocol-adapter-example/pom.xml]: IMHO this should be Java 11

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/DemoTCPApplication.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
+package org.eclipse.hono.example.protocoladapter;
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/DemoTCPApplication.java]: please add the copyright/license header

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/ServiceProperties.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package org.eclipse.hono.example.protocoladapter;
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/ServiceProperties.java]: please add the copyright/license header

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/adapter/CommandAndControlReceiver.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
+package org.eclipse.hono.example.protocoladapter.adapter;
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/adapter/CommandAndControlReceiver.java]: please add the copyright/license header

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/adapter/TelemetryAndEventSender.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
+package org.eclipse.hono.example.protocoladapter.adapter;
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/adapter/TelemetryAndEventSender.java]: please add the copyright/license header

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/controller/ProtocolAdapterExample.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
+package org.eclipse.hono.example.protocoladapter.controller;
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/controller/ProtocolAdapterExample.java]: please add the copyright/license header

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/interfaces/ICommandHandler.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package org.eclipse.hono.example.protocoladapter.interfaces;
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/interfaces/ICommandHandler.java]: please add the copyright/license header

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/interfaces/ICommandHandler.java [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+public interface ICommandHandler {
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/interfaces/ICommandHandler.java]: we do not use a special prefix character for interfaces

File: example/protocol-adapter-example/src/main/resources/application.yml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+spring:
+  profiles:
+    active: dev
+
+app:
+  serverPort: 6666
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/resources/application.yml]: missing EOL

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/interfaces/ICommandHandler.java [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+     * Pass through function to handle commands and return response body
+     *
+     * @param commandPayload body of command
+     * @param subject        subject of command
+     * @param contentType    HTML content type
+     * @param isOneWay       signals if response string necessary
+     * @return
+     */
+    String handleCommand(String commandPayload, String subject, String contentType, boolean isOneWay);
+}
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/interfaces/ICommandHandler.java]: missing EOL

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/controller/ProtocolAdapterExample.java [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+    /**
+     * Start listening for commands
+     * <p>
+     * Connection properties have to be set with {@link #setAMQPClientProps(String, int, String, String, ICommandHandler) } beforehand
+     */
+    public void listenCommands() {
+        commandAndControlReceiver.listenCommands();
+    }
+
+}
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/controller/ProtocolAdapterExample.java]: missing EOL

File: site/homepage/content/sandbox.md [hunk 17]
```diff
@@ -39,9 +39,12 @@ All services are exposed via the same ports as used in the guide.
...
   can be used to consume telemetry/event messages from the sandbox as below:
-  
-        java -jar hono-cli-*-exec.jar --hono.client.host=hono.eclipse.org --hono.client.port=15671 --hono.client.tlsEnabled=true --hono.client.username=consumer@HONO --hono.client.password=verysecret --spring.profiles.active=receiver`
-  Note that  only the *receiver* profile is activated but not the *ssl* profile.
+
+  ~~~sh
+  # in directory where the hono-cli-*-exec.jar file has been downloaded to
+  java -jar hono-cli-*-exec.jar --hono.client.host=hono.eclipseprojects.io --hono.client.port=15671 --hono.client.tlsEnabled=true --hono.client.username=consumer@HONO --hono.client.password=verysecret --spring.profiles.active=receiver
+  ~~~
+  Note that only the *receiver* profile is activated but not the *ssl* profile.
```
Comment (sophokles73) [inline site/homepage/content/sandbox.md]: this doesn't seem related to your PR, is it?

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/adapter/CommandAndControlReceiver.java [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+    public void setAMQPClientProps(final String host, final int port, final String username, final String password, final ICommandHandler commandHandler) {
+        final ClientConfigProperties props = new ClientConfigProperties();
+        props.setHost(host);
+        props.setPort(port);
+        props.setUsername(username);
+        props.setPassword(password);
+        setClientConfig(props);
+        this.commandHandler = commandHandler;
+    }
+}
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/adapter/CommandAndControlReceiver.java]: missing EOL

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/ServiceProperties.java [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+     *
+     * @return The Vert.x instance.
+     */
+    @Bean
+    public Vertx vertx() {
+        final VertxOptions options = new VertxOptions().setWarningExceptionTime(1500000000);
+        return Vertx.vertx(options);
+    }
+
+}
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/ServiceProperties.java]: missing EOL

File: example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/DemoTCPApplication.java [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+     * Connection properties have to be set with {@link #initConnection(String, int, String, String) } beforehand
+     *
+     * @param payload        message payload
+     * @param messageAddress address can be either "telemetry" or "event"
+     * @return
+     */
+    public Future<String> sendAMQPMessage(String payload, String messageAddress) {
+        return protocolAdapterExample.sendAMQPMessage(payload, messageAddress);
+    }
+}
```
Comment (sophokles73) [inline example/protocol-adapter-example/src/main/java/org/eclipse/hono/example/protocoladapter/DemoTCPApplication.java]: missing EOL

File: example/protocol-adapter-example/README.md (hunk 2 above)
Comment (ko-j) [reply example/protocol-adapter-example/README.md]: done

File: example/protocol-adapter-example/pom.xml (hunk 6 above)
Comment (ko-j) [reply example/protocol-adapter-example/pom.xml]: properties removed, since inherited by parent.

File: example/protocol-adapter-example/pom.xml [hunk 21]
```diff
@@ -1,22 +1,34 @@
 <?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline example/protocol-adapter-example/pom.xml]: since this is a new file, it should only contain the current year

File: example/protocol-adapter-example/scripts/create_hono_device.sh [hunk 22]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+#*******************************************************************************
+# Copyright (c) 2016, 2020 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline example/protocol-adapter-example/scripts/create_hono_device.sh]: current year only
this probably applies to all the other files you have added as well ...

File: example/protocol-adapter-example/pom.xml [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+   
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.eclipse.hono</groupId>
+        <artifactId>hono-bom</artifactId>
```
Comment (sophokles73) [inline example/protocol-adapter-example/pom.xml]: the parent of this module is `hono-example`

File: example/hono-client-examples/pom.xml [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+   
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.eclipse.hono</groupId>
+    <artifactId>hono-bom</artifactId>
```
Comment (sophokles73) [inline example/hono-client-examples/pom.xml]: the parent of this module now is `hono-example`

File: site/homepage/content/sandbox.md (hunk 17 above)
Comment (ko-j) [reply site/homepage/content/sandbox.md]: That's right. Was caused by a merger mishap. Removed

**Comment by ko-j** [comment]
PR now contains the changes concerning the latest proposals.

File: example/pom.xml [hunk 25]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
-    Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
+    Copyright (c) 2020 Contributors to the Eclipse Foundation
```
[the remark this answers is not in the document]
Comment (ko-j) [reply example/pom.xml]: Since the original file in this directory was moved to [`example/hono-client-examples/pom.xml`](https://github.com/eclipse/hono/pull/1724/files#diff-926b5f64866f1096050ab4477635ec36) one could argue that this is a new file.
